src/combinatorics/ssyt.lean

```diff
@@ -0,0 +1,102 @@
+/-
+Copyright (c) 2022 Jake Levinson. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Jake Levinson
+-/
+import combinatorics.young_diagram
+
+/-!
+# Semistandard Young tableaux
+
+A semistandard Young tableau is a filling of a Young diagram by natural numbers, such that
+the entries are weakly increasing left-to-right along rows (i.e. for fixed `i`), and
+strictly-increasing top-to-bottom along columns (i.e. for fixed `j`).
+
+An example of an SSYT of shape `μ = [4, 2, 1]` is:
+
+```text
+0 0 0 2
+1 1
+2
+```
+
+We represent an SSYT as a function `ℕ → ℕ → ℕ`, which is required to be zero for all pairs
+`(i, j) ∉ μ` and to satisfy the row-weak and column-strict conditions on `μ`.
+
+
+## Main definitions
+
+- `ssyt (μ : young_diagram)` : semistandard Young tableaux of shape μ
+- `ssyt.highest_weight (μ : young_diagram)`: the semistandard Young tableau whose `i`th row
+  consists entirely of `i`s, for each `i`.
+
+## Tags
+
+Semistandard Young tableau
+
+## References
+
+<https://en.wikipedia.org/wiki/Young_tableau>
+
+-/
+
+/-- A semistandard Young tableau (SSYT) is a filling of the cells of a Young diagram by natural
+numbers, such that the entries in each row are weakly increasing (left to right), and the entries
+in each column are strictly increasing (top to bottom).
+
+Here, an SSYT is represented as an unrestricted function `ℕ → ℕ → ℕ`, but which is required to
+vanish outside `μ`. --/
+structure ssyt (μ : young_diagram) :=
+  (entry : ℕ → ℕ → ℕ)
+  (row_weak : ∀ {i j1 j2 : ℕ}, j1 < j2 → (i, j2) ∈ μ → entry i j1 ≤ entry i j2)
+  (col_strict : ∀ {i1 i2 j: ℕ}, i1 < i2 → (i2, j) ∈ μ → entry i1 j < entry i2 j)
+  (zeros' : ∀ {i j}, (i, j) ∉ μ → entry i j = 0)
+
+namespace ssyt
+
+instance fun_like {μ : young_diagram} : fun_like (ssyt μ) ℕ (λ _, ℕ → ℕ) :=
+{ coe := λ T, T.entry,
```

Comment:
I believe this should work:
```suggestion
{ coe := ssyt.entry,
```
I *believe* that this saves a beta reduction, but this might be a pointless suggestion.